src/lib/y2storage/storage_class_wrapper.rb

```diff
@@ -0,0 +1,258 @@
+  # more Ruby-oriented API and adding extra methods as needed.
+  #
+  # A class in the Y2Storage namespace can include this mixing and then use
+  # the wrap_class macro to point to the Storage class to wrap. Then it can
+  # use the storage_forward and storage_class_forward macros to define methods
+  # and class methods. The calls to those methods will be automatically
+  # forwarded to the wrapped object. The wrapped object will be specified when
+  # creating the object (the initializer is also provided by this mixin).
+  module StorageClassWrapper
+    def self.included(base)
+      base.extend(ClassMethods)
+    end
+
+    # Constructor for the wrapper object
+    #
+    # @param the Storage object to wrap
+    def initialize(object)
+      cast_method = :"to_#{self.class.storage_class_underscored_name}"
+      @storage_object =
+        if Storage.respond_to?(cast_method)
+          Storage.send(cast_method, object)
+        else
+          object
+        end
+    end
+
+    # Equivalent to this object in the Storage world, i.e. the wrapped object.
+    def to_storage_value
+      @storage_object
+    end
+
+    # Class methods to be added
+    module ClassMethods
+      # Macro to define the class in the Storage namespace to be wrapped.
+      #
+      # Since libstorage enforces explicit type check and explicit downcast for
+      # every object, this macro is also used to specify all the wrapper classes
+      # to which it makes sense to automatically downcast the objects (i.e. the
+      # existing subclasses for this in the Y2Storage namespace).
+      #
+      # @param storage_class [Class] class to wrap
+      # @param downcast_to [String, Array<String>] name(s) of the existing
+      #   subclasses in Y2Storage to automatically perform downcasts to. Do not
+      #   include the "Y2Storage" namespace in the class name (i.e. names
+      #   relatively qualified).
+      #
+      # @example Basic usage of wrap_class
+      #
+      #   module Y2Storage
+      #     class BlkDevice
+      #       include StorageClassWrapper
+      #       wrap_class Storage::BlkDevice, downcast_to: ["Disk", "MyPartition"]
+      #     end
+      #
+      #     class Disk < BlkDevice
+      #       include StorageClassWrapper
+      #       wrap_class Storage::Disk
+      #     end
+      #
+      #     class MyPartition < BlkDevice
+      #       include StorageClassWrapper
+      #       wrap_class Storage::Partition
+      #     end
+      #   end
+      #
+      #   blk_device = Storage::find_by_name(a_devicegraph, "/dev/sda2")
+      #   blk_device.class #=> Storage::BlkDevice
+      #   # The user should never need to call this constructor
+      #   y2_device = Y2Storage::BlkDevice.downcasted_new(blk_device)
+      #   y2_device.class #=> Y2Storage::Partition
+      #
+      def wrap_class(storage_class, downcast_to: nil)
+        @storage_class = storage_class
+        @downcast_class_names = Array(downcast_to).compact
```

Comment:
no, single element is fine and valid approach for me. What I do not like is default `nil` instead of `[]` which makes more sense for me.
Doc I mean was that types are `[String, Array<String>]` without `nil` which is usually indicator for me to not pass nil there otherwise it crash.